Sources/Plasma/CoreLib/hsDarwin.h

```diff
@@ -0,0 +1,105 @@
+Copyright (C) 2011  Cyan Worlds, Inc.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#ifndef _hsDarwin_inc_
+#define _hsDarwin_inc_
+
+#include <string_theory/string>
+#include <string_theory/format>
+
+#ifdef HS_BUILD_FOR_APPLE
+#include <CoreFoundation/CoreFoundation.h>
+
+#ifdef __OBJC__
+    @class NSString;
+#else
+    class NSString;
+#endif
+
+[[nodiscard]]
+#if __has_feature(attribute_cf_returns_retained)
+__attribute__((cf_returns_retained))
+#endif
+inline CFStringRef CFStringCreateWithSTString(const ST::string& str)
+{
+    return CFStringCreateWithBytes(kCFAllocatorDefault, (const UInt8*)str.data(), str.size(), kCFStringEncodingUTF8, false);
+}
+
+inline ST::string STStringFromCFString(CFStringRef str, ST::utf_validation_t validation = ST_DEFAULT_VALIDATION)
+{
+    CFRange range = CFRangeMake(0, CFStringGetLength(str));
+    CFIndex strBufSz = 0;
+    CFStringGetBytes(str, range, kCFStringEncodingUTF8, 0, false, nullptr, 0, &strBufSz);
+    ST::char_buffer buffer;
+    buffer.allocate(strBufSz);
+    CFStringGetBytes(str, range, kCFStringEncodingUTF8, 0, false, (UInt8*)buffer.data(), strBufSz, nullptr);
+
+    return ST::string(buffer, validation);
+}
+
+#define ST2CFString(str) CFStringCreateWithSTString(str)
+#define CF2STString(str) STStringFromCFString((CFStringRef)str)
```

Comment:
It might be good to remove these shorthands. They mask the ownership rules. (The word create is pretty critical in the above function to understanding the memory ownership relationship.)